src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -131,33 +122,43 @@ where s.Value.SagaData.GetType() == sagaType && (s.Key != saga.Id)
 
         ConcurrentDictionary<Guid, VersionedSagaEntity> data = new ConcurrentDictionary<Guid, VersionedSagaEntity>();
 
-        int version;
-
         class VersionedSagaEntity
         {
             public VersionedSagaEntity(IContainSagaData sagaData, VersionedSagaEntity original = null)
             {
                 SagaData = DeepClone(sagaData);
-                versionCache = original?.versionCache ?? new ConditionalWeakTable<IContainSagaData, SagaVersion>();
+                if (original != null)
+                {
+                    original.ConcurrencyCheck(sagaData);
+
+                    versionCache = original.versionCache;
+                    version = original.version;
+                    version++;
```

Comment:
I think @timbussmann is referring to the fact that the delegates in the concurrent dictionary can end up being called multiple times but only one wins...

> If you call AddOrUpdate simultaneously on different threads, addValueFactory may be called multiple times, but its key/value pair might not be added to the dictionary for every call.